proxy/http/HttpSessionManager.cc

```diff
@@ -146,52 +146,75 @@ ServerSessionPool::acquireSession(sockaddr const *addr, CryptoHash const &hostna
+          (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_CERT) || validate_cert(sm, netvc))) {
+        // Check if on the same thread
+        if (ethread == netvc->thread) {
+          Debug("http_ss", "Session found on the same thread");
+          to_return = iter;
+          break;
+        } else if (nullptr == to_return) {
+          to_return = iter;
+        }
       }
-      ++first;
+      ++iter;
     }
-    if (zret == HSM_DONE) {
-      to_return = first;
+    if (nullptr != to_return) {
+      zret = HSM_DONE;
       if (!to_return->is_multiplexing()) {
         this->removeSession(to_return);
       }
-    } else if (first != m_fqdn_pool.end()) {
+    } else if (iter != m_fqdn_pool.end()) {
       Debug("http_ss", "Failed find entry due to name mismatch %s", sm->t_state.current.server->name);
     }
   } else if (TS_SERVER_SESSION_SHARING_MATCH_MASK_IP & match_style) { // matching is not disabled.
-    auto first = m_ip_pool.find(addr);
-    // The range is all that is needed in the match IP case, otherwise need to scan for matching fqdn
-    // And matches the other constraints as well
-    // Note the port is matched as part of the address key so it doesn't need to be checked again.
+    auto iter = m_ip_pool.find(addr);
+    // The range is all that is needed in the match IP case,
+    // otherwise need to scan for matching fqdn and matches
+    // the other constraints as well.
+    // Note the port is matched as part of the address key
+    // so it doesn't need to be checked again.
+    // Prefer a session that's already on this thread.
     if (match_style & (~TS_SERVER_SESSION_SHARING_MATCH_MASK_IP)) {
-      while (first != m_ip_pool.end() && ats_ip_addr_port_eq(first->get_remote_addr(), addr)) {
-        if ((!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTONLY) || first->hostname_hash == hostname_hash) &&
-            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_SNI) || validate_sni(sm, first->get_netvc())) &&
-            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTSNISYNC) || validate_host_sni(sm, first->get_netvc())) &&
-            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_CERT) || validate_cert(sm, first->get_netvc()))) {
-          zret = HSM_DONE;
-          break;
+      while (iter != m_ip_pool.end() && ats_ip_addr_port_eq(iter->get_remote_addr(), addr)) {
+        auto *const netvc = iter->get_netvc();
+        if ((!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTONLY) || iter->hostname_hash == hostname_hash) &&
+            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_SNI) || validate_sni(sm, netvc)) &&
+            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTSNISYNC) || validate_host_sni(sm, netvc)) &&
+            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_CERT) || validate_cert(sm, netvc))) {
+          // Check if on same thread
+          if (netvc->thread == ethread) {
```

Comment:
How is this only triggered in the global locked case?

Reply:
it's not this one is actually a change to the pools in general.  if necessary i could split this into 2 prs.

Reply:
done and tested.